File: 0739-daily-temperatures/0739-daily-temperatures.java
class Pair {
    int temperature;
    int index;

    public Pair(int temperature, int index) {
        this.temperature = temperature;
        this.index = index;
    }
}

class Solution {
    public int[] dailyTemperatures(int[] temperatures) {
        int n = temperatures.length;
        int[] answer = new int[n];
        Stack<Pair> stack = new Stack<>();

        for (int i = 0; i < n; i++) {
            if (!stack.isEmpty() && stack.peek().temperature < temperatures[i]) {
                while (!stack.isEmpty() && stack.peek().temperature < temperatures[i]) {
                    Pair now = stack.pop();
                    answer[now.index] = i - now.index;
                }
            }
            stack.push(new Pair(temperatures[i], i));
        }

        return answer;
    }
}

File: 0739-daily-temperatures/0739-daily-temperatures.py
class Solution:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        n = len(temperatures)
        answer = [0 for i in range(n)]
        stack = list()

        for i in range(n):
            if stack and stack[-1][0] < temperatures[i]:
                while stack and stack[-1][0] < temperatures[i]:
                    t, d = stack.pop()
                    answer[d] = i - d
            stack.append((temperatures[i], i))

        return answer